src/nvim/liveupdate.c

```diff
@@ -0,0 +1,167 @@
+  linedata.items = xcalloc(sizeof(Object), line_count);
+  for (size_t i = 0; i < line_count; i++) {
+    linenr_T lnum = 1 + (linenr_T)i;
+
+    const char *bufstr = (char *)ml_get_buf(buf, lnum, false);
+    Object str = STRING_OBJ(cstr_to_string(bufstr));
+
+    // Vim represents NULs as NLs, but this may confuse clients.
+    strchrsub(str.data.string.data, '\n', '\0');
+
+    linedata.items[i] = str;
+  }
+
+  Array args = ARRAY_DICT_INIT;
+  args.size = 3;
+  args.items = xcalloc(sizeof(Object), args.size);
+
+  // the first argument is always the buffer number
+  args.items[0] = INTEGER_OBJ(buf->handle);
+  args.items[1] = ARRAY_OBJ(linedata);
+  args.items[2] = BOOLEAN_OBJ(false);
+
+  channel_send_event(channel_id, "LiveUpdateStart", args);
+  return true;
+}
+
+void liveupdate_send_end(buf_T *buf, uint64_t channelid) {
+    Array args = ARRAY_DICT_INIT;
+    args.size = 1;
+    args.items = xcalloc(sizeof(Object), args.size);
+    args.items[0] = INTEGER_OBJ(buf->handle);
+    channel_send_event(channelid, "LiveUpdateEnd", args);
+}
+
+void liveupdate_unregister(buf_T *buf, uint64_t channelid) {
+  size_t size = kv_size(buf->liveupdate_channels);
+  if (!size) {
+    return;
+  }
+
+  // go through list backwards and remove the channel id each time it appears
+  // (it should never appear more than once)
+  size_t j = 0;
+  size_t found = 0;
+  for (size_t i = 0; i < size; i++) {
+    if (kv_A(buf->liveupdate_channels, i) == channelid) {
+      found++;
+    } else {
+      // copy item backwards into prior slot if needed
+      if (i != j) {
+        kv_A(buf->liveupdate_channels, j) = kv_A(buf->liveupdate_channels, i);
+      }
+      j++;
+    }
+  }
+
+  if (found) {
+    // remove X items from the end of the array
+    buf->liveupdate_channels.size -= found;
+
+    // make a new copy of the active array without the channelid in it
+    liveupdate_send_end(buf, channelid);
+  }
+}
+
+void liveupdate_unregister_all(buf_T *buf) {
+  size_t size = kv_size(buf->liveupdate_channels);
+  if (size) {
+    for (size_t i = 0; i < size; i++) {
+      liveupdate_send_end(buf, kv_A(buf->liveupdate_channels, i));
+    }
+    kv_destroy(buf->liveupdate_channels);
+    kv_init(buf->liveupdate_channels);
+  }
+}
+
+void liveupdate_send_changes(buf_T *buf, linenr_T firstline, int64_t num_added,
+                             int64_t num_removed) {
+  // if one the channels doesn't work, put its ID here so we can remove it later
+  uint64_t badchannelid = 0;
+
+  // notify each of the active channels
+  size_t size = kv_size(buf->liveupdate_channels);
+  if (size) {
+    for (size_t i = 0; i < size; i++) {
+      uint64_t channelid = kv_A(buf->liveupdate_channels, i);
+
+      // send through the changes now channel contents now
+      Array args = ARRAY_DICT_INIT;
+      args.size = 4;
+      args.items = xcalloc(sizeof(Object), args.size);
+
+      // the first argument is always the buffer number
+      args.items[0] = INTEGER_OBJ(buf->handle);
```

Comment:
Buffer should be sent using EXT encoding.  See https://gitter.im/neovim/neovim?at=581609f85a1cfa016e61e784